src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,758 @@
+    FUNCTION_LOG_OBJECT_FORMAT(value, walFilterToLog, buffer, bufferSize)
+
+static inline
+void
+checkOutputSize(Buffer *const output, const size_t size)
+{
+    if (bufRemains(output) <= size)
+    {
+        bufResize(output, bufSize(output) + size);
+    }
+}
+
+// Reading the next page from the input buffer. If the input buffer is exhausted, remember the current step and returns true.
+// In this case, we should exit the process function to get a new input buffer.
+static inline
+bool
+readPage(WalFilterState *const this, const Buffer *const input, const int step)
+{
+    if (this->input_offset >= bufUsed(input))
+    {
+        this->current_step = step;
+        this->input_offset = 0;
+        this->same_input = false;
+        return true;
+    }
+    this->page = bufPtrConst(input) + this->input_offset;
+    this->input_offset += XLOG_BLCKSZ;
+    this->page_offset = 0;
+    this->current_step = 0;
+    this->current_header = (XLogPageHeaderData *) this->page;
+    this->page_offset += MAXALIGN(XLogPageHeaderSize(this->current_header));
+
+    // Make sure that WAL belongs to supported Greenplum version, since magic value is different in different versions.
+    if (this->current_header->xlp_magic != this->walInterface->header_magic)
+    {
+        THROW_FMT(FormatError, "wrong page magic");
+    }
+
+    lstAdd(this->headers, this->current_header);
+
+    return false;
+}
+
+static inline uint32_t
+getRecordSize(const unsigned char *const buffer)
+{
+    return ((XLogRecord *) (buffer))->xl_tot_len;
+}
+
+static bool
+readRecord(WalFilterState *const this, const Buffer *const input)
+{
+    // Go back to the place where the reading was interrupted due to the exhaustion of the input buffer.
+    switch (this->current_step)
+    {
+        case 1:
+            goto step1;
+
+        case 2:
+            goto step2;
+
+        case 3:
+            goto step3;
+    }
+
+    if (this->page_offset == 0 || this->page_offset >= XLOG_BLCKSZ)
+    {
+step1:
+        if (readPage(this, input, 1))
+        {
+            return false;
+        }
+
+        if (this->is_begin)
+        {
+            this->is_begin = false;
+            // There may be an unfinished record from the previous file at the beginning of the file. Just skip it.
+            // We should have handled it elsewhere.
+            if (this->current_header->xlp_info & XLP_FIRST_IS_CONTRECORD &&
+                !(this->current_header->xlp_info & XLP_FIRST_IS_OVERWRITE_CONTRECORD))
+            {
+                this->page_offset += MAXALIGN(this->current_header->xlp_rem_len);
+            }
+        }
+        else if (this->current_header->xlp_info & XLP_FIRST_IS_CONTRECORD)
+        {
+            THROW_FMT(FormatError, "should not be XLP_FIRST_IS_CONTRECORD");
+        }
+    }
+
+    // Record header can be split between pages but first field xl_tot_len is always on single page
+    this->record_size = getRecordSize(this->page + this->page_offset);
```

Comment:
may be eliminate getRecordSize function?
```suggestion
    XLogRecord *record = this->page + this->page_offset;
    this->record_size = record->xl_tot_len;
```

Reply:
I made the `record_size` variable local since it is only used for memory allocation. But making the `record` variable local seems like a bad idea to me. Since it will be confusing with the field in `this`. And it is impossible to make `record` completely local, since in the process of reading we can exit the function.